src/main/java/util/ApiQuotaManager.java

```diff
@@ -0,0 +1,169 @@
+                API_QUOTA_BUFFER,
+                DEFAULT_REFRESH_PERIOD_IN_MINS);
+        this.refreshTimer.restartTimer((int) Utility.minsToSecs(refreshDurationInMinutes));
+        logger.info("Refresh period updated to " + refreshDurationInMinutes
+                + "mins with API calls used in previous refresh cycle is " + apiCallsUsedInPreviousRefresh
+                + ", current API quota is " + info.getRemainingRequests() + " and next API quota top-up in "
+                + info.getNextRefreshInMinutesFromNow() + "mins.");
+    }
+
+    /**
+     * Computes the API calls used in previous refresh.
+     * @param remainingRequests The number of API requests remaining in the current API quota window.
+     * @return The number of API calls used in previous refresh.
+     */
+    private int computeApiCallsUsedInPreviousRefresh(int remainingRequests) {
+        int difference = previousRemainingApiRequests - remainingRequests;
+        previousRemainingApiRequests = remainingRequests;
+
+        if (difference >= 0) {
+            return difference;
+        }
+        return apiCallsUsedInPreviousRefresh;
+    }
+
+    /**
+     * Computes the TickerTimer period that is used for refreshing the issues periodically
+     * Assumes future refreshes will take the same number of API calls as the last refresh and find out the refresh
+     * duration that will spread out the refreshes until the next API quota top up.
+     *
+     * For some cases where the computed refresh rate is equal to the remainingTimeInMins,
+     * bufferTime is added to ensure that the next refresh happens after the apiQuota renewal.
+     *
+     * PMD is suppressed to allow explicit parenthesis.
+     *
+     * @param apiQuota The remaining allowed API requests until the next API quota renewal.
+     *                 Pre-condition: >= 0
+     * @param remainingTimeInMins The remaining time left until the next API quota renewal.
+     *                           Pre-condition: >= 0
+     * @param apiCallsUsedInPreviousRefresh The amount of API used in the previous API refresh.
+     *                                      Pre-condition: >= 0
+     * @param apiQuotaBuffer The amount of API calls that is set aside for manual operations by the user.
+     *                       Pre-condition: >= 0
+     * @param minRefreshPeriod The minimal refresh period that will be used.
+     *                         There are 3 conditions that this will be used.
+     *                         1) During application initialisation, when apiCallsUsedInPreviousRefresh is zero.
+     *                         2) When computed refresh rate is < minRefreshPeriod.
+     *                         3) When remainingTimeInMins = 0. This indicates that refreshes can occur in the
+     *                            nearest time possible.
+     *                            Recommended value : ApiQuotaManager.DEFAULT_REFRESH_PERIOD_IN_MINS
+     *                         Pre-condition: > 0
```

Comment:
You can move the three conditions under the `@return`, because these are the conditions under which the return value will be the minRefreshPeriod instead of the computed value.